fe_rtos/src/fe_alloc.rs

```diff
@@ -64,6 +61,25 @@ unsafe impl GlobalAlloc for KernelAllocator {
     }
 }
 
+// Returns the appropriate bit mask for a bit position with a certain number of
+// bits remaining to process. Also updates cur_block by the number of blocks/bits
+// included in the returned bit mask.
+fn get_bit_mask(bit: usize, blocks_remaining: usize, cur_block: &mut usize) -> u8 {
+    // Look up table for potential bit masks
+    const BIT_LUT: [u8; 8] = [0x01, 0x03, 0x07, 0x0F, 0x1F, 0x3F, 0x7F, 0xFF];
+
+    //If we need to include more bits than are represented in
+    //this byte of the free list, set the rest of the bits,
+    //otherwise, just set the bits you need.
+    if blocks_remaining >= 8 - bit {
+        *cur_block += 8 - bit;
+        BIT_LUT[8 - bit - 1] << bit
+    } else {
+        *cur_block += blocks_remaining;
+        BIT_LUT[blocks_remaining - 1] << bit
+    }
+}
+
 pub(crate) unsafe fn alloc(layout: LayoutFFI) -> *mut u8 {
     static mut ADDING_TO_LIST: bool = false;
 
```

Comment:
Should the `(HEAP as usize) == 0` check and subsequent call to `init_heap()` be protected by `ALLOC_LOCK`? I think as is, you could end up where two calls to alloc get queued up and start initializing the heap at the same time. Requiring `ALLOC_LOCK` to be held when calling `init_heap()` would mean you could take those calls out of `init_heap()` as well.

Reply:
You're right. I think this has worked well in the past because the allocator has been used before scheduling has started in our example projects, but there might be some edge cases where this is not that case. I'll make the change.